[#1336] Demo extension (#1353)
open-formulieren/open-forms: docs/developers/extending.rst

.. _developers_extending:

=========
Extending
=========

Open Forms ("the backend") is implemented with a pluggable architecture.

Within Open Forms there are modules which focus on specific functionality. In these modules, there are plugins that
implement the required functionality using multiple vendors. At the time of writing, the following modules exist:

* Registrations
* Authentication
* Prefill
* Payments
* Appointments
* Formio formatters
* Formio custom types
* Validations

Modules and plugins are Python packages with some Django bindings. They are called

> by the Open Forms core functionality (forms, submissions...).

Reviewer: This is a lot of repetition from `developers/architecture.rst`, so I'd reference that instead. Also, the technial documentation about that needs to be expanded upon, but that's another ticket.

We support two ways to add new plugins:

#. Adding new plugins directly in Open Forms. These are extensions that serve the greater community and can be
   contributed as a pull request to Open Forms (more information can be found in the `CONTRIBUTING.md`_).

#. Adding 'external' plugins. These are extensions that are extremely specific or can't be open-sourced.
   Instructions on how to implement such 'external plugins' are described below.

.. _CONTRIBUTING.md: https://github.com/open-formulieren/open-forms/blob/master/CONTRIBUTING.md

Creating an external plugin
===========================

The `demo extension <https://github.com/open-formulieren/demo-extension>`_ is an example of an extension plugin.

> An extension must be implement as a Django package, which is a stricter form of a

Reviewer: ```suggestion
An extension must be implemented as a Django package, which is a stricter form of a
```

.. _default-app: https://github.com/maykinmedia/default-app

Configurable options
^^^^^^^^^^^^^^^^^^^^

Extensions cannot require modifications to the Django settings of Open Forms. Any run-time configuration option can
be specified as:

#. Deployment time environment variables. See `plugin.py:35 <https://github.com/open-formulieren/demo-extension/blob/main/demo_extension/plugin.py#L35>`_.

#. Dynamic options using ``django-solo``, with the advantage that configuration can be
   modified at runtime through the Admin interface. See `models.py`_ and `admin.py`_ in the demo plugin.
   If you add a solo model, you will have to generate migrations.
   See the :ref:`section below<test_extension_plugin_locally>` for details.

#. For certain types of plugins, like for registration backends, it is also possible to add form-specific options.
   These can be specified with a serializer. See `config.py`_ for details.

If any initialization or module loading is needed, this should be done as part of the
extension's ``extension.apps.ExtensionConfig.ready`` hook, which is called when Django
bootstraps. This is the mechanism by which Open Forms plugins are registered

> (refer to the :ref:`plugin section<plugins_index>` for further details).

Reviewer: this looks copied a bit too-verbatim from the issue ;)

Author: :sweat_smile:

.. _config.py: https://github.com/open-formulieren/demo-extension/blob/main/demo_extension/config.py

.. _test_extension_plugin_locally:

Testing locally
---------------

In order to test your external plugin locally, you can use symlinks. Inside the ``src/`` directory of Open Forms,
create a symlink to your package. For example, in the case of the demo extension package:

.. code-block:: bash

    ln -s /path/to/plugin/demo_extension .

Now add an environment variable ``OPEN_FORMS_EXTENSIONS`` with the name of your package. The variable is a
comma-separated list of valid python identifiers (i.e. the python package names). For example:

.. code-block:: bash

    export OPEN_FORMS_EXTENSIONS=demo_extension

The Open Forms ``base.py`` settings module will extract this configuration value and
append the list of packages to ``INSTALLED_APPS``, which causes them to be loaded by
Django.

If you need to generate migrations for your package, you can now do it as follows (from within the Open Forms directory):

.. code-block:: bash

    python src/manage.py makemigrations demo_extension
    python src/manage.py migrate

> If you created a solo model, you can add the configuration page to the Admin. To do this, login into the Open Forms

Reviewer: ```suggestion
If you created a solo model, you can add the configuration page to the Admin. To do this, log in into the Open Forms
```
?

Admin:

#. Go to the **Configuratie** > **Application groups**.

#. Click on **Configuratie**.

#. In the **Models** section, look for the name of your configuration model in the left table (for the demo extension, this was  ``demo_extension.Demoextensionconfig``).
   Then double click on it to add it to the right table.

> #. Save the configuration.

Reviewer: We'll have to find a better solution for this, as I think the post-migrate signal handlers deletes any manual changes and restores the fixture configuration.

=========
Extending
=========

Open Forms ("the backend") is implemented with a :ref:`pluggable architecture<plugins_index>`.
Modules and plugins are Python packages with some Django bindings. They get called
by the Open Forms core functionality (forms, submissions...).

We support two ways to add new plugins:

#. Adding new plugins directly in Open Forms. These are extensions that serve the greater community and can be
   contributed as a pull request to Open Forms (more information can be found in the `CONTRIBUTING.md`_).

#. Adding 'external' plugins. These are extensions that are extremely specific or can't be open-sourced.
   Instructions on how to implement such 'external plugins' are described below.

.. _CONTRIBUTING.md: https://github.com/open-formulieren/open-forms/blob/master/CONTRIBUTING.md

Creating an external plugin
===========================

An extension must be implemented as a Django package, which is a stricter form of a
regular Python package. They can be developed independently from the open-formulieren Github organisation.

The extensions can be loaded at deployment time through the environment variable ``OPEN_FORMS_EXTENSIONS`` which

> specifies the Python name of the extensions to load. The Open Forms ``base.py`` settings module will extract this
> configuration value and append the list of packages to ``INSTALLED_APPS``, which causes them to be loaded by
> Django. For an extension to be loaded, it needs to be present in the ``PYTHONPATH``.

Reviewer: ```suggestion
specifies the Python name of the extensions to load. Open Forms will extract this configuration value and 
load the referenced extension packages. For an extension to be loaded, it needs to be present in the ``PYTHONPATH``.
```
Simplified a bit without mentioning too many implementation details.

This means that it needs to be either:

* In the ``src/`` directory of Open Forms. It is then automatically picked up.

> * Anywhere on the file path, but the ``PYTHONPATH`` is modified to include the path to the extension.

Reviewer: ```suggestion
* Anywhere on the file path, but the ``PYTHONPATH`` environment variable is modified to include the path to the extension.
```

* In the relevant site-packages directory, similarly to when an Open Forms dependency is installed with ``pip install``
   and ``virtualenv``.

> To create a Docker image that extends an Open Forms release, a custom ``Dockerfile`` can be created and the extension
> source code can be added to the desired location. Another option is to mount the source code as a volume in a custom
> ``docker-compose.yml`` file.

Reviewer: ```suggestion
Building and distributing the extended Open Forms
-------------------------------------------------
You can build a custom Docker image extending an Open Forms release with a custom `Dockerfile`, in which the
extension source code is added to the desired location. Another option is to mount the source code as a volume 
in a custom ``docker-compose.yml`` file or as part of your Kubernetes manifests, depending on how you deploy your instance(s).
```
